django/api/viewsets/upload.py

```diff
@@ -54,92 +55,61 @@ def datasets_list(self, request):
-        records_inserted_msg = "{} records inserted. This table currently contains {} records.".format(records_inserted, final_count)
-        if done != True:
-            try:
-                error_location = done[1]
-                error = done[0]
-                error_row = 0
-                error_msg = "There was an error. Please check your file and ensure you have the correctly named worksheets, column names, and data types in cells and reupload. Error: {}".format(error)
-                if len(done) > 2:
-                    error_row = done[2]
-                error_type = type(error).__name__
-                field_names = [f.name for f in model._meta.fields]   
-                if error_location == 'data':
-                    if error_type in (type(LookupError), type(KeyError), 'KeyError') :
-                        error_msg = "Please make sure you've uploaded a file with the correct data including the correctly named columns. There was an error finding: {}. This dataset requires the following columns: {}".format(error, field_names)
-                    elif error_type == 'ValueError' or type(ValueError):
-                        ## note for next batch of scripts, possibly add str(type(ValueError)) 
-                        ## to this but check for impacts to other exceptions
-                        error_msg = "{} on row {}. Please make sure you've uploaded a file with the correct data.".format(error, error_row)
-                    elif isinstance(error, ValidationError):
-                        error_msg ="Issue with cell value on row {}. {}".format(error_row, str(error)[2:-2])
-                elif error_location == 'file':
-                    error_msg = "{}. Please make sure you've uploaded a file with the correct data including the correctly named worksheets.".format(error)
-                if error_msg[-1] != '.':
-                    error_msg+='.'
-                error_msg += records_inserted_msg
-                return Response(error_msg, status=status.HTTP_400_BAD_REQUEST)
-            except Exception as error:
-                print(error)
-                return Response('There was an issue!', status=status.HTTP_400_BAD_REQUEST)
-        else:
-            return Response(records_inserted_msg, status=status.HTTP_201_CREATED)
+
+            config = DATASET_CONFIG.get(dataset_selected)
+            if not config:
+                return Response(f"Dataset '{dataset_selected}' is not supported.", status=status.HTTP_400_BAD_REQUEST)
+            model = config['model']
+            columns = config.get('columns')
+            mapping = config.get('column_mapping')
+            sheet_name = config.get('sheet_name', 'Sheet1')  # Default to 'Sheet1' if not specified
+            preparation_functions = config.get('preparation_functions', [])
+            validation_functions = config.get('validation_functions', [])
+
+            if replace_data:
+                model.objects.all().delete()
```

Comment:
Looking good! Just one little thing: Instead of doing this here, I would pass the "replace_data" variable to the import_from_xls() function, and then to the load_data() function, and do the conditional deletion there so that it's part of the same transaction as the bulk_create. That way, if there's an error with the bulk_create, the deletion gets rolled back.